commit ff3893ca8d3a11560024e6a068c45ce59aa77f88 (refs/changes/37/2137/5)
Author: Dave Eddy <dave@daveeddy.com>
Date:   2017-06-26T15:37:06-04:00 (2 years, 3 months ago)
    
    OS-6204 async.forEachSeries RangeError with many inputs
    Reviewed by: Josh Wilsdon <josh@wilsdon.ca>
    Approved by: Josh Wilsdon <josh@wilsdon.ca>

diff --git a/src/vm/sbin/vmadmd.js b/src/vm/sbin/vmadmd.js
index 4edae44d..734445f2 100755
--- a/src/vm/sbin/vmadmd.js
+++ b/src/vm/sbin/vmadmd.js
@@ -21,7 +21,7 @@
  *
  * CDDL HEADER END
  *
- * Copyright (c) 2016, Joyent, Inc. All rights reserved.
+ * Copyright 2017 Joyent, Inc.
  *
  */
 
@@ -2428,8 +2428,17 @@ function main()
                 'zpool'
             ];
 
-            VM.lookup({}, {fields: lookup_fields}, function (e, vmobjs) {
-                async.forEachSeries(vmobjs, function (obj, upg_cb) {
+            VM.lookup({}, {fields: lookup_fields},
+                function vmLookup(e, vmobjs) {
+
+                if (e) {
+                    log.error({err: e}, 'VM.lookup failed');
+                    process.exit(2);
+                }
+                vasync.forEachPipeline({
+                    inputs: vmobjs,
+                    func: function upgradeSingleVM(obj, upg_cb) {
+
                     function finishUpgrade(vmobj) {
                         if (!seen_vms.hasOwnProperty(vmobj.zonename)) {
                             seen_vms[vmobj.zonename] = {
@@ -2579,6 +2588,14 @@ function main()
                             });
                         });
                     });
+
+                    }
+                }, function upgradeVMsComplete(_err, results) {
+                    if (_err) {
+                        log.error({err: _err}, 'Error processing all VMs');
+                    } else {
+                        log.debug('Finished processing all VMs');
+                    }
                 });
             });
         });
